File: convex/offers.ts
import { action, internalQuery, mutation, query } from "./_generated/server";
import { v } from "convex/values";
import { api, internal } from "./_generated/api";
import { Doc, Id } from "./_generated/dataModel";

// Helper type for internal vector search results
type VectorSearchResult = {
  _id: Id<"offers">;
  _score: number;
};

// List offers with optional filters
export const list = query({
  args: {
    status: v.optional(v.string()),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    if (args.status) {
      return await ctx.db
        .query("offers")
        .withIndex("by_status", (idx) => idx.eq("status", args.status!))
        .order("desc")
        .take(args.limit ?? 100);
    }
    return await ctx.db
      .query("offers")
      .order("desc")
      .take(args.limit ?? 100);
  },
});

// Get offer by ID
export const getById = query({
  args: { id: v.string() },
  handler: async (ctx, args) => {
    try {
      // Try to parse as Convex Id first
      const docId = ctx.db.normalizeId("offers", args.id);
      if (!docId) return null;
      return await ctx.db.get(docId);
    } catch (e) {
      return null;
    }
  },
});

export const get = getById;

// Internal get for actions
export const getInternal = internalQuery({
  args: { id: v.id("offers") },
  handler: async (ctx, args) => {
    return await ctx.db.get(args.id);
  },
});

// Count active offers
export const countActive = query({
  args: {},
  handler: async (ctx) => {
    const active = await ctx.db
      .query("offers")
      .withIndex("by_status", (q) => q.eq("status", "ACTIVE"))
      .collect();
    return active.length;
  },
});

// Create offer
export const create = mutation({
  args: {
    sourceName: v.optional(v.string()),
    sourcePhone: v.string(),
    sourceGroup: v.string(),
    medication: v.string(),
    medicationRaw: v.string(),
    quantity: v.optional(v.number()),
    unit: v.optional(v.string()),
    price: v.optional(v.number()),
    currency: v.optional(v.string()),
    status: v.string(),
    urgencyLevel: v.string(),
    aiConfidence: v.number(),
    notes: v.optional(v.string()),
    rawMessageId: v.string(),
    contentEmbedding: v.optional(v.array(v.number())),
    expiryDate: v.optional(v.string()),
    batchNumber: v.optional(v.string()),
    expiryInfo: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    const now = Date.now();
    return await ctx.db.insert("offers", {
      ...args,
      createdAt: now,
      updatedAt: now,
    });
  },
});

// Update offer status
export const updateStatus = mutation({
  args: {
    id: v.string(),
    status: v.string(),
  },
  handler: async (ctx, args) => {
    const docId = args.id as Id<"offers">;
    await ctx.db.patch(docId, {
      status: args.status,
      updatedAt: Date.now(),
    });
    return await ctx.db.get(docId);
  },
});

// Find recent duplicate
export const findRecentDuplicate = query({
  args: {
    senderPhone: v.string(),
    medication: v.string(),
    withinMs: v.number(),
  },
  handler: async (ctx, args) => {
    const cutoff = Date.now() - args.withinMs;
    return await ctx.db
      .query("offers")
      .withIndex("by_source", (q) => q.eq("sourcePhone", args.senderPhone))
      .filter((q) =>
        q.and(
          q.eq(q.field("medication"), args.medication),
          q.eq(q.field("status"), "ACTIVE"),
          q.gte(q.field("createdAt"), cutoff)
        )
      )
      .first();
  },
});

// Search offers by medication (LIKE %query% approximate)
export const search = query({
  args: {
    query: v.string(),
    limit: v.optional(v.number()),
    status: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    if (args.status) {
      return await ctx.db
        .query("offers")
        .withSearchIndex("search_medication", (q) =>
          q.search("medication", args.query).eq("status", args.status!)
        )
        .take(args.limit ?? 100);
    }

    return await ctx.db
      .query("offers")
      .withSearchIndex("search_medication", (q) =>
        q.search("medication", args.query)
      )
      .take(args.limit ?? 100);
  },
});

// Offer with score
type OfferWithScore = Doc<"offers"> & { _score: number };

// Vector search for semantic duplicates
export const searchSimilar = action({
  args: {
    embedding: v.array(v.number()),
    limit: v.optional(v.number()),
    statusFilter: v.optional(v.string()),
    withinMs: v.optional(v.number()),
    similarityThreshold: v.optional(v.number()),
  },
  handler: async (ctx, args): Promise<OfferWithScore[]> => {
    const results: VectorSearchResult[] = await ctx.vectorSearch(
      "offers",
      "by_embedding",
      {
        vector: args.embedding,
        limit: args.limit ?? 10,
        filter: args.statusFilter
          ? (q) => q.eq("status", args.statusFilter!)
          : undefined,
      }
    );

    const cutoff = args.withinMs ? Date.now() - args.withinMs : 0;
    const offersWithScores: OfferWithScore[] = [];

    for (const result of results) {
      const offer = await ctx.runQuery(api.offers.getById, {
        id: result._id,
      });
      if (offer) {
        if (
          (!args.withinMs || offer.createdAt >= cutoff) &&
          (!args.similarityThreshold ||
            result._score >= args.similarityThreshold)
        ) {
          offersWithScores.push({ ...offer, _score: result._score });
        }
      }
    }

    return offersWithScores;
  },
});

// Delete offers created before cutoff (matches SeaORM delete_before)
export const deleteBefore = mutation({
  args: { cutoff: v.number() },
  handler: async (ctx, args) => {
    const toDelete = await ctx.db
      .query("offers")
      .withIndex("by_created", (q) => q.lt("createdAt", args.cutoff))
      .collect();

    for (const doc of toDelete) {
      await ctx.db.delete(doc._id);
    }
    return toDelete.length;
  },
});

File: core/crates/db/src/convex/offers.rs
//! Convex OfferRepository implementation

use std::sync::Arc;

use async_trait::async_trait;
use chrono::{DateTime, Utc};

use super::client::ConvexClient;
use crate::Result;
use crate::convex_args;
use crate::params::{FindDuplicateParams, SemanticDuplicateParams};
use crate::traits::{ItemStatus, OfferModel, OfferRepository};

/// Convex-backed offer repository
pub struct ConvexOfferRepo {
    client: Arc<ConvexClient>,
}

impl ConvexOfferRepo {
    pub fn new(client: Arc<ConvexClient>) -> Self {
        Self { client }
    }
}

#[async_trait]
impl OfferRepository for ConvexOfferRepo {
    async fn get_by_id(&self, id: &str) -> Result<Option<OfferModel>> {
        self.client
            .query("offers:get", convex_args! { "id" => id })
            .await
    }

    async fn get_active(&self, limit: i64, _offset: i64) -> Result<Vec<OfferModel>> {
        self.client
            .query(
                "offers:list",
                convex_args! {
                    "status" => "ACTIVE",
                    "limit" => limit
                },
            )
            .await
    }

    async fn search(&self, query: &str, limit: i64, _offset: i64) -> Result<Vec<OfferModel>> {
        self.client
            .query(
                "offers:search",
                convex_args! {
                    "query" => query,
                    "limit" => limit,
                    "status" => "ACTIVE"
                },
            )
            .await
    }

    async fn count_active(&self) -> Result<i64> {
        self.client
            .query("offers:countActive", convex_args! {})
            .await
    }

    async fn find_recent_duplicate(
        &self,
        params: FindDuplicateParams<'_>,
    ) -> Result<Option<OfferModel>> {
        self.client
            .query(
                "offers:findRecentDuplicate",
                convex_args! {
                    "senderPhone" => params.sender_phone,
                    "medication" => params.medication,
                    "withinMs" => params.within.num_milliseconds()
                },
            )
            .await
    }

    async fn save(&self, offer: &OfferModel) -> Result<OfferModel> {
        let price_f64 = offer.price.as_ref().and_then(|p| {
            use rust_decimal::prelude::ToPrimitive;
            p.to_f64()
        });
        let quantity_f64 = offer.quantity.as_ref().and_then(|q| {
            use rust_decimal::prelude::ToPrimitive;
            q.to_f64()
        });
        let expiry_str = offer.expiry_date.map(|d| d.to_string());

        let id: String = self
            .client
            .mutation(
                "offers:create",
                convex_args! {
                    "rawMessageId" => &offer.raw_message_id,
                    "sourcePhone" => &offer.source_phone,
                    "sourceName" => offer.source_name.as_ref(),
                    "sourceGroup" => &offer.source_group,
                    "medication" => &offer.medication,
                    "medicationRaw" => &offer.medication_raw,
                    "quantity" => quantity_f64,
                    "unit" => offer.unit.as_ref(),
                    "price" => price_f64,
                    "currency" => offer.currency.as_ref(),
                    "expiryDate" => expiry_str,
                    "batchNumber" => offer.batch_number.as_ref(),
                    "notes" => offer.notes.as_ref(),
                    "status" => format!("{:?}", offer.status),
                    "urgencyLevel" => format!("{:?}", offer.urgency_level),
                    "expiryInfo" => offer.expiry_info.as_ref(),
                    "aiConfidence" => offer.ai_confidence,
                    "contentEmbedding" => offer.content_embedding.as_ref().map(|v| v.to_vec())
                },
            )
            .await?;

        let mut saved = offer.clone();
        saved.id = id;
        Ok(saved)
    }

    async fn update_status(&self, id: &str, status: ItemStatus) -> Result<OfferModel> {
        self.client
            .mutation(
                "offers:updateStatus",
                convex_args! {
                    "id" => id,
                    "status" => format!("{:?}", status)
                },
            )
            .await
    }

    async fn find_semantic_duplicates(
        &self,
        params: SemanticDuplicateParams<'_>,
    ) -> Result<Vec<OfferModel>> {
        // Convert f32 embedding to f64 for Convex
        let embedding_vec: Vec<f64> = params.embedding.iter().map(|x| *x as f64).collect();

        self.client
            .action(
                "offers:searchSimilar",
                convex_args! {
                    "embedding" => embedding_vec,
                    "limit" => 10i64,
                    "statusFilter" => "ACTIVE",
                    "withinMs" => params.within.num_milliseconds(),
                    "similarityThreshold" => params.similarity_threshold
                },
            )
            .await
    }

    async fn delete_before(&self, cutoff: &DateTime<Utc>) -> Result<u64> {
        self.client
            .mutation(
                "offers:deleteBefore",
                convex_args! { "cutoff" => cutoff.timestamp_millis() },
            )
            .await
    }
}
